x-pack/legacy/plugins/siem/server/client/factory.ts

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Legacy } from 'kibana';
+
+import { KibanaRequest } from '../../../../../../src/core/server';
+import { SiemClient } from './client';
+
+interface SetupDependencies {
+  getSpaceId?: (request: KibanaRequest) => string | undefined;
+  config: Legacy.Server['config'];
+}
+
+export class SiemClientFactory {
+  private getSpaceId!: SetupDependencies['getSpaceId'];
+  private config!: SetupDependencies['config'];
+
+  public setup({ getSpaceId, config }: SetupDependencies) {
+    this.getSpaceId = getSpaceId;
+    this.config = config;
+  }
+
+  public create(request: KibanaRequest): SiemClient {
+    const spaceId = this.getSpaceId?.(request) ?? 'default';
+    return new SiemClient(spaceId, this.config);
+  }
+}
```

Comment:
So with classes and objects comes the two part phases where you sometimes cannot call something like `create` until after `setup` has been called such as in this example.

I wouldn't deactivate TypeScript's type assertion in these two areas:

```ts
private getSpaceId!: SetupDependencies['getSpaceId'];
private config!: SetupDependencies['config'];
```

As it's trying to do you a favor and let you know that you actually could accidentally call `create` before `setup` has been called. It makes sense.

Instead though you could do something like this where it gives you back safety like TypeScript is kind of wanting:

```ts
export class SiemClientFactory {
  private getSpaceId: SetupDependencies['getSpaceId'] | undefined;
  private config: SetupDependencies['config'] | undefined;

  public setup({ getSpaceId, config }: SetupDependencies) {
    this.getSpaceId = getSpaceId;
    this.config = config;
  }

  public create(request: KibanaRequest): SiemClient {
    if (this.config == null) {
      throw new TypeError(
        'Cannot call create until config is non-null which means setup should have been called'
      );
    }
    const spaceId = this.getSpaceId?.(request) ?? 'default';
    return new SiemClient(spaceId, this.config);
  }
}
```

And it makes it to where if someone accidentally forgets to call `create` before calling `setup` they are going to get an exception immediately rather than having this accepting a undefined/null when it does not have the typing for that potential value:

```
return new SiemClient(spaceId, this.config);
```

Reply:
Unit test for the condition above if you change it:

```ts
    it('cannot call create without calling setup first', () => {
      const factory = new SiemClientFactory();
      const mockRequest = httpServerMock.createKibanaRequest();
      expect(() => factory.create(mockRequest)).toThrow(
        'Cannot call create until config is non-null which means setup should have been called'
      );
    });
```

Thanks for adding the unit tests so I can put them here for code reviews btw! 😎

Reply:
https://github.com/elastic/kibana/pull/58292/commits/01c9a7b886bc31544efbbc3764a4cd50f2c1fb34